root/edit/details/AddArtist.js

```diff
@@ -0,0 +1,168 @@
+/*
+ * @flow
+ * Copyright (C) 2019 MetaBrainz Foundation
+ *
+ * This file is part of MusicBrainz, the open internet music database,
+ * and is licensed under the GPL version 2, or (at your option) any
+ * later version: http://www.gnu.org/licenses/gpl-2.0.txt
+ */
+
+import React from 'react';
+
+import {
+  artistBeginAreaLabel,
+  artistBeginLabel,
+  artistEndAreaLabel,
+  artistEndLabel,
+} from '../../artist/utils';
+import {addColon, l} from '../../static/scripts/common/i18n';
+import {lp_attributes} from '../../static/scripts/common/i18n/attributes';
+import formatDate from '../../static/scripts/common/utility/formatDate';
+import isDateEmpty from '../../static/scripts/common/utility/isDateEmpty';
+import DescriptiveLink from '../../static/scripts/common/components/DescriptiveLink';
+import EntityLink from '../../static/scripts/common/components/EntityLink';
+import yesNo from '../../static/scripts/common/utility/yesNo';
+import formatIsni from '../../utility/formatIsni';
+
+type AddArtistEditT = {|
+  ...EditT,
+  +display_data: {|
+    ...CommentRoleT,
+    ...DatePeriodRoleT,
+    +area: AreaT | null,
+    +artist: AreaT,
+    +begin_area: AreaT | null,
+    +end_area: AreaT | null,
+    +gender: GenderT | null,
+    +ipi_codes: $ReadOnlyArray<string>,
+    +isni_codes: $ReadOnlyArray<string>,
+    +name: string,
+    +sort_name: string,
+    +type: ArtistTypeT,
+  |},
+|};
+
+const AddArtist = ({edit}: {edit: AddArtistEditT}) => {
+  const display = edit.display_data;
+
+  return (
+    <>
+      <table className="details">
+        <tr>
+          <th>{addColon(l('Artist'))}</th>
+          <td>
+            <EntityLink
+              entity={display.artist}
+            />
+          </td>
+        </tr>
+      </table>
+
+      <table className="details add-artist">
+        <tr>
+          <th>{addColon(l('Name'))}</th>
+          <td>{display.name}</td>
+        </tr>
+
+        <tr>
+          <th>{addColon(l('Sort name'))}</th>
+          <td>{display.sort_name}</td>
+        </tr>
+
+        {display.comment ? (
+          <tr>
+            <th>{addColon(l('Disambiguation'))}</th>
+            <td>{display.comment}</td>
+          </tr>
+        ) : null}
+
+        {display.type ? (
+          <tr>
+            <th>{addColon(l('Type'))}</th>
+            <td>{lp_attributes(display.type.name, 'artist_type')}</td>
+          </tr>
+        ) : null}
+
+        {display.gender && display.gender.name ? (
+          <tr>
+            <th>{addColon(l('Gender'))}</th>
+            {/* $FlowFixMe */}
+            <td>{lp_attributes(display.gender.name, 'gender')}</td>
```

Comment:
Flow refuses to understand if this is inside && display.gender.name then it doesn't matter there's no .name in gender = null.

Although actually, gender is not really null, it's either GenderT or `''`. Not sure if that should be changed or this should be changed or we should leave it be since it works anyway.

Reply:
Set `const gender = display.gender;` above and reference `gender` here, then the error should go away. Flow is pretty conservative with refinements, so if it sees a function call like `addColon` that it can't prove doesn't modify `display`, it'll invalidate the refinement. Saving it to the variable works around that, since it can't be modified.

I'd greatly prefer accurate types even if we don't think it makes a difference, since I trust the computer more than I to know if it does. :)